src/main/scala/internal/Graph.scala

```diff
@@ -0,0 +1,425 @@
+ * made up of three types of vertices, Nodes, Sources, and Sinks, with edges between them.
+ *
+ * Nodes consist of a unique id, and an Operator. Each Node also has zero, one, or multiple ordered dependencies
+ * on Nodes and Sources elsewhere in the graph. These dependencies are represented as incoming edges in the DAG,
+ * with data flowing into the node. Dependency order matters, as the incoming data is passed directly to the
+ * Operator with each dependency as a separate parameter. Operators may take zero, one, or multiple ordered
+ * input parameters, and always produce a single output.
+ *
+ * Next, we have Sources, a special type of vertex. A source is an incomplete input connection coming in from the
+ * outside world. A workload cannot be executed without providing information about what data should be passed into
+ * each Source. Sources are also treated by utility methods as endpoints of the DAG, useful for specifying how to
+ * connect multiple workload DAGs together. Each Source has a unique id, and Sources always have zero
+ * dependencies/incoming edges.
+ *
+ * Finally, we have Sinks, another special type of vertex. Sinks each have a unique id, and exactly one dependency
+ * on either a source or a node. Sinks are output endpoints exposed to the outside world. They are used as markers
+ * during execution to specify what results of executing a workload to output. Like Sources, they are also referenced
+ * by utilities to to connect workloads together. Each Sink has a unique id, and exactly one dependency on either
+ * a Node or a Source.
+ *
+ * @param sources  The set of all [[SourceId]]s of sources in the graph
+ * @param sinkDependencies  A map of [[SinkId]] to the id of the node or source the sink depends on
+ * @param operators  A map of [[NodeId]] to the operator contained within that node
+ * @param dependencies  A map of [[NodeId]] to the node's ordered dependencies
+ */
+case class Graph(
+    sources: Set[SourceId],
+    sinkDependencies: Map[SinkId, NodeOrSourceId],
+    operators: Map[NodeId, Operator],
+    dependencies: Map[NodeId, Seq[NodeOrSourceId]]
+  ) {
+
+  /**
+   * Get the set of ids of all nodes in this graph.
+   */
+  def nodes: Set[NodeId] = operators.keySet
+
+  /**
+   * Get the set of ids of all sinks in this graph.
+   */
+  def sinks: Set[SinkId] = sinkDependencies.keySet
+
+  /**
+   * Get the dependencies of a given node in this graph.
+   */
+  def getDependencies(id: NodeId): Seq[NodeOrSourceId] = dependencies(id)
+
+  /**
+   * Get the dependency of a given sink in this graph.
+   */
+  def getSinkDependency(id: SinkId): NodeOrSourceId = sinkDependencies(id)
+
+  /**
+   * Get the operator at a given node in this graph.
+   */
+  def getOperator(id: NodeId): Operator = operators(id)
+
+  /**
+   * Get {@param num} [[NodeId]]s that don't clash with those of existing nodes.
+   */
+  private def nextNodeIds(num: Int): Seq[NodeId] = {
+    val maxId = (nodes.map(_.id) + 0).max
```

Comment:
Actually I'll just add an if statement instead, that'll be more straightforward.